pkg/controller/catalogsource/catalogsource_controller.go

```diff
@@ -0,0 +1,138 @@
+			if _, ok := defaultCatalogsources[e.Meta.GetName()]; ok {
+				// If DeleteStateUnknown is true it implies that the Delete event was missed
+				// and we can ignore it.
+				if e.DeleteStateUnknown {
+					return false
+				}
+				return true
+			}
+			return false
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			if _, ok := defaultCatalogsources[e.Meta.GetName()]; ok {
+				return true
+			}
+			return false
+		},
+	}
+
+	err = c.Watch(&source.Kind{Type: &olm.CatalogSource{}}, &handler.EnqueueRequestForObject{}, pred)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileOperatorHub implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileCatalogSource{}
+
+// ReconcileCatalogSource reconciles a CatalogSource object
+type ReconcileCatalogSource struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+}
+
+func (r *ReconcileCatalogSource) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log.Infof("Reconciling default CatalogSource %s", request.Name)
+
+	_, defaultCatalogsources := defaults.GetGlobalDefinitions()
+	defaultCatsrcDef := defaultCatalogsources[request.Name]
+	// Fetch the CatalogSource instance
+	instance := &olm.CatalogSource{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			if !operatorhub.GetSingleton().Get()[defaultCatsrcDef.Name] {
+				createNewCatsrcInstance(r.client, defaultCatsrcDef)
+			}
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	if instance.DeletionTimestamp != nil {
+		if !operatorhub.GetSingleton().Get()[defaultCatsrcDef.Name] {
+			createNewCatsrcInstance(r.client, defaultCatsrcDef)
+		}
+	}
+
+	if !defaults.AreCatsrcSpecsEqual(&defaultCatsrcDef.Spec, &instance.Spec) {
+		if err := r.client.Delete(context.TODO(), instance); err != nil {
+			log.Warnf("Could not set default CatalogSource %s's spec back to desired default state. Error in deleting updated CatalogSource: %s", defaultCatsrcDef.GetName(), err.Error())
+		}
+		if !operatorhub.GetSingleton().Get()[defaultCatsrcDef.Name] {
+			if err := createNewCatsrcInstance(r.client, defaultCatsrcDef); err != nil {
+				log.Warnf("Could not set default CatalogSource %s's spec back to desired default state. Error in creating new CatalogSource: %s", defaultCatsrcDef.GetName(), err.Error())
+			}
+		}
+		if err = r.client.Update(context.TODO(), &defaultCatsrcDef); err != nil {
+
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func createNewCatsrcInstance(client client.Client, catsrc olm.CatalogSource) error {
+	err := client.Create(context.TODO(), &catsrc)
+	if err != nil {
+		log.Warnf("Could not recreate default CatalogSource %s. Error: %s", catsrc.GetName(), err.Error())
```

Comment:
nit: I think we should return the `err` after the log and return nil below.

Reply:
Done.